src/Papercut.Network/SmtpCommands/AuthSmtpCommand.cs

```diff
@@ -0,0 +1,173 @@
+            {
+                try
+                {
+                    switch (args[0]?.ToUpperInvariant())
+                    {
+                        case "PLAIN":
+                            RunAuthPlain(args);
+                            break;
+
+                        case "LOGIN":
+                            RunAuthLogin();
+                            break;
+
+                        default:
+                            result = Connection.SendLine("504 Authentication method not implemented");
+                            break;
+                    }
+                }
+                finally
+                {
+                    if (result == null)
+                    {
+                        result = (!string.IsNullOrEmpty(Session?.Username)
+                               ? Connection.SendLine("235 Authentication successful")
+                               : Connection.SendLine("535 Authentication failed"));
+                    }
+                }
+            }
+            else result = Connection.SendLine("501 Authentication method not provided");
+
+            result.Wait();    
+        }
+
+        private void RunAuthPlain(string[] args)
+        {
+            // AUTH PLAIN accepts a base64-encoded authentication token comprised of
+            // an authorisation value, username and password, separated by the NUL char.
+            // The token is supplied either on the same line as the AUTH PLAIN command
+            // or on a separate line, after being prompted by a 334 message
+
+            string authToken = null;
+
+            if (args.Length != 1)
+            {
+                // auth token is sent on same line as AUTH PLAIN
+                authToken = args[1];
+            }
+            else
+            {
+                // auth token is to be sent separately on the next line
+                authToken = Connection.Client.ReadTextStream(reader => {
+                    Connection.SendLine("334").Wait();
+                    return reader.ReadLine();
+                });
+            }
+
+            if (!string.IsNullOrEmpty(authToken))
+            {
+                // decode the token as Base64, then split on \0 (nul);         
+                var decodedToken = DecodeString(authToken)?.Split('\0');
+
+                // [0] is the authorisation name (a SASL thing we probably don't need)
+                // [1] is the username; [2] is the password
+                if ((decodedToken?.Length ?? 0) == 3)
+                {
+                    Authenticate(decodedToken[1], decodedToken[2]);
+                }
+            }
+        }
+
+        private void RunAuthLogin()
+        {
+            // AUTH LOGIN prompts for two base64-encoded strings -- the username
+            // and password -- by sending two consecutive 334 messages.  As a foible,
+            // the two prompts are also base64-encoded.  We pre-encode these for some
+            // tiny micro-optimisation
+
```

Comment:
I feel like, in this situation, "micro-optimization" isn't beneficial -- as it's most likely less clear for other developers.

Reply:
The alternatives I can see here are:
 * to send something like `$"334 {Convert.ToBase64String(Encoding.Utf8.GetBytes("Username:"))}"` each time on the fly
 * for `usernamePrompt` and `passwordPrompt` not to be `const` so the assignment can be made each time, based on the `Convert` call
 * to uplift `usernamePrompt` and `passwordPrompt` to `private static readonly` class members, so the assignment can be done once, based on the `Convert` call

Do you have a preference/alternative?

Reply:
How about a ToBase64() extension? You can make it const/static -- but out of curiosity I benchmarked the call (using the excellent BenchmarkDotNet) and it's 75ns (or 0.000075ms). Not worth optimizing giving it will only be called a few times per connection.

Reply:
OK, I've added a `ToBase64String` extension to `GeneralExtensions` (in keeping with `AsString`) and folded the encoding directly into the `SendLine` calls.